### DISCUSSION TITLE
Is there a way for the currently running procedure to know if it’s the last?

### DISCUSSION THREAD

**Discussion by Did-Mas** [pymeasure/pymeasure#1184] [Q&A] [answered]
I am trying to implement functionality to modify a procedure’s behavior depending on whether it is the last one executed. However, I’m unsure how to pass this information to the procedure.

From what I’ve observed, the program retrieves this information by calling the *has_next()* function on the experiment queue in the *Manager* class. I thought it might be possible to approach this similarly to how the *should_stop()* function was implemented, but the *Worker* class does not have access to the *Manager* class attributes.

Is there an effective way to pass this information?

**Comment by BenediktBurger** [comment]
@CasperSchippers , I think this question is for you.

**Comment by CasperSchippers** [comment, marked as answer]
That is a very good question.

Currently, there is no convenient way to acquire this information, but I think it might be possible, indeed somewhat similar to how the `should_stop` method, but rather how it is _monkeypatched_ (rather than implemented): the `should_stop` has an abstract definition in the `Procedure` class and is monkeypatched by the  `Worker`.

For the `has_next` method, there are two steps needed: 1) the manager needs to monkeypatch the `has_next` method on the `Worker` when an instance is created (from the `ExperimentQueue`, `self.experiments`), and the `Worker` in turn needs to monkeypatch it on the the `Procedure`. Then I think you should be able to access that information by calling `self.has_next()` in the `Procedure`.

Hope this helps. I do think this would be a useful feature, so do feel free to open a PR with this addition if you get it working.

**Reply by Did-Mas** [reply]
Thank you for your reply; this solution makes sense. The only thing I’m still unsure about is the function name from a Procedure perspective. Calling it “has_next” seems confusing since the procedure doesn’t "have" a next one. A more intuitive name might be “is_last,” for example, but that would require a negation somewhere along the way. But maybe I’m overthinking it. ;)

**Reply by BenediktBurger** [reply]
Pull request is #1185
